parser/parser.go

```diff
@@ -577,6 +582,29 @@ func (p *parser) endAmbiguity() {
 	}
 }
 
+type parserProgress struct {
+	initialized bool
+	offset      int
+}
+
+// checkProgress checks that the parser has made progress since it was called last with this parserProgress.
+func (p *parser) checkProgress(progress *parserProgress) bool {
+	parserOffset := p.current.StartPos.Offset
+
+	if !progress.initialized {
+		progress.initialized = true
+		progress.offset = parserOffset
+		return true
+	}
```

Comment:
Could we do the initialization when the `parserProgress` was created? So that it doesn't need to check it here for every iteration..

Reply:
That's a good idea, yeah. I was trying to make the "zero value" of the type usable, so that we wouldn't have to ensure it is initialized (e.g. forget to call a constructor function), but that might be more efficient than checking for initialization on each call 👍

Reply:
Improved this in c5ac940fa176fa8a3ceead949d5fdd01a2bb66e8